tests/symtable.test.ts

```diff
@@ -0,0 +1,281 @@
+    assert(!spam.lookup("x").is_namespace());
+
+    assert(top.lookup("spam").get_namespace() === spam);
+    const ns_test = top.lookup("namespace_test");
+    assertEquals(ns_test.get_namespaces()?.length, 2);
+    assertThrows(() => ns_test.get_namespace());
+
+    const ns_test_2 = top.lookup("glob");
+    assertEquals(ns_test_2.get_namespaces()?.length, 0);
+    assertThrows(() => ns_test_2.get_namespace());
+});
+
+Deno.test("test_assigned", () => {
+    assert(spam.lookup("x").is_assigned());
+    assert(spam.lookup("bar").is_assigned());
+    assert(top.lookup("spam").is_assigned());
+    assert(Mine.lookup("a_method").is_assigned());
+    assert(!internal.lookup("x").is_assigned());
+});
+
+Deno.test("test_annotated", () => {
+    const st1 = buildSymbolTable(runParserFromString("def f():\n    x: int\n"), "test", "exec").top!;
+    const st2 = st1.get_children()[0];
+    assert(st2.lookup("x").is_local());
+    assert(st2.lookup("x").is_annotated());
+    assert(!st2.lookup("x").is_global());
+    const st3 = buildSymbolTable(runParserFromString("def f():\n    x = 1\n"), "test", "exec").top!;
+    const st4 = st3.get_children()[0];
+    assert(st4.lookup("x").is_local());
+    assert(!st4.lookup("x").is_annotated());
+
+    // Test that annotations in the global scope are valid after the
+    // variable is declared as nonlocal.
+    const st5 = buildSymbolTable(runParserFromString("global x\nx: int"), "test", "exec").top!;
+    assert(st5.lookup("x").is_global());
+
+    // Test that annotations for nonlocals are valid after the
+    // variable is declared as nonlocal.
+    const st6 = buildSymbolTable(
+        runParserFromString("def g():\n" + "    x = 2\n" + "    def f():\n" + "        nonlocal x\n" + "    x: int"),
+        "test",
+        "exec"
+    );
+});
+
+Deno.test("test_imported", () => {
+    assert(top.lookup("sys").is_imported());
+});
+
+Deno.test("test_name", () => {
+    assertEquals(top.get_name(), "top");
+    assertEquals(spam.get_name(), "spam");
+    assertEquals(spam.lookup("x").get_name(), "x");
+    assertEquals(Mine.get_name(), "Mine");
+});
+
+Deno.test("test_class_info", () => {
+    assertEquals(Mine.get_methods(), ["a_method"]);
+});
+
+Deno.test("test_filename_correct", () => {
+    // Bug tickler: SyntaxError file name correct whether error raised
+    // while parsing or building symbol table.
+    function checkfilename(brokencode: string, offset: number) {
+        try {
+            buildSymbolTable(runParserFromString(brokencode, "exec", "spam"), "spam", "exec");
+        } catch (e) {
+            assertEquals(e.traceback[0], "spam");
+            assertEquals(e.traceback[1], 1);
+            assertEquals(e.traceback[2], offset);
+            return;
+        }
+
+        assert(false, "no SyntaxError for '" + brokencode + "'");
+    }
+    checkfilename("def f(x): foo)(", 14); // parse-time
+    checkfilename("def f(x): global x", 10); // symtable-build-time
```

Comment:
I had to cheat here, in cpython this is 11 instead of 10 in our ast it's 10 as well so 🤷

Reply:
https://github.com/python/cpython/blob/3.10/Python/symtable.c#L1373-L1377

Looks like they just add 1 to the offset.

Reply:
haha ok thanks 😄